parsers/BB.py

```diff
@@ -0,0 +1,181 @@
+    zone_key,
+    session: Session,
+    target_datetime: datetime | None,
+    logger: Logger,
+) -> pd.DataFrame:
+    session = session_with_auth(session)
+
+    params = {"name": "operational-data", "$sort[created_at]": -1}
+
+    header = None
+    table = None
+
+    if target_datetime is None:
+        data = session.get(base_url + "/barbados/uploads", params=params).json()
+        header, table = parse_operational_data(data)
+    elif target_datetime > datetime(year=2023, month=2, day=27):
+        # WHEN HISTORICAL DATA IS AVAILABLE
+        # convert target datetime to local datetime
+
+        expected_date = target_datetime.astimezone(
+            ZoneInfo("America/Barbados")
+        ).strftime("%Y-%m-%d")
+
+        for i in range(10):
+            url_date = (
+                (target_datetime + timedelta(days=i))
+                .astimezone(ZoneInfo("America/Barbados"))
+                .strftime("%Y-%m-%d")
+            )
+            params["created_at[$lt]"] = url_date
+            data = session.get(base_url + "/barbados/uploads", params=params).json()
+            header, table = parse_operational_data(data)
+            if table[0][0][:10] == expected_date:
+                break
+            if table[0][0][:10] > expected_date:
+                raise ParserException("BB.py", "Date not found", zone_key)
+    else:
+        # WHEN HISTORICAL DATA IS NOT AVAILABLE
+        raise ParserException(
+            "BB.py",
+            "This parser is not yet able to parse dates before 2000-01-01",
+            zone_key,
+        )
+
+    if table is None or header is None:
+        raise ParserException("BB.py", "No data found", zone_key)
+
+    operational_data = pd.DataFrame(
+        table,
+        columns=header,
+    )
+    columns = [
+        "24 hr monitored solar production",
+        "24 hr monitored wind production",
+        "24 hr storage charging",
+        "24 hr storage discharging",
+        "24 hr fossil production",
+        "24 hr Net demand",
+        "24 hr Curtailment",
+    ]
+    for columm in columns:
+        operational_data[columm] = pd.to_numeric(operational_data[columm])
+    return operational_data
+
+
+def fetch_production(
+    zone_key: ZoneKey = ZoneKey("BB"),
+    session: Session = Session(),
+    target_datetime: datetime | None = None,
+    logger: Logger = getLogger(__name__),
+) -> list[dict] | dict:
+    operational_data = fetch_operational_data(
+        zone_key, session, target_datetime, logger
+    )
+
+    production_list = ProductionBreakdownList(logger=logger)
+
+    for _index, row in operational_data.iterrows():
+        production_list.append(
+            zoneKey=zone_key,
+            datetime=datetime.fromisoformat(row["datetime"]),
+            production=ProductionMix(
+                solar=row["24 hr monitored solar production"],
+                wind=row["24 hr monitored wind production"],
+                unknown=row["24 hr fossil production"],
```

Comment:
Not sure what to map fossil production to.

Reply:
Hi!
I've found this info in a PDF in the energy.gov.bb webpage:

![Screenshot_20240217_174656](https://github.com/electricitymaps/electricitymaps-contrib/assets/84802276/0adcc2c2-038b-49d4-91f9-ca602e35f528)

The data, albeit dated, shows thermal is oil. "Bagasse" is a bit confusing as it seems to not be reported on the API, or to be aggregated with oil.

Reply:
https://dataservices.acelerex.com/barbados/grid-operation/generation/long-term-forecast shows, "Long Term Forecast" tab shows they currently don't have, or at least didn't forecast having, biomass or landfill gas capacity.

Maybe it doesn't exist anymore?

Reply:
Barbados IRENA profile shows no bioenergy capacity. Maybe there is some plant that uses mixed fuel? Or maybe some factory burns bagasse for self consumption?

Reply:
I tried looking for context. This document which I found by googling for ["Fuel Inputs for Electricity Production 2015 Barbados"](https://faolex.fao.org/docs/pdf/bar199547.pdf) says 

>Over the last 20 years, the fuel feedstock used in Barbados for the production of electricity
consisted in large measure of fossil fuels. The main fuels included heavy fuel oil, kerosene,
diesel and natural gas. Within the sugar industry, electricity production was mainly fuelled
by bagasse.


Elsewhere it says:

>Also, in light of a declining sugar cane industry on the island, there is increasing confidence
in the agricultural sector that there can be greater use of biomass and biogas in the
energy sector. Bagasse has, at times, been used to generate energy in the sugar cane
industry and there is considerable interest in building on this experience.

It's quite possible that bagasse is no longer used, or that it's only used in captive power plants that aren't grid-connected.

Reply:
I also noticed in the ["Barbados National Energy Policy (BNEP) 2019 - 2030"](https://energy.gov.bb/download/national-energy-policy-2019-2030/), probably where you found the diagram, that the values in the diagram are rounded. One page up (page 42), it has the percentages with one decimal point, showing, among others, that a bit of natural gas was also used for electricity generation in 2015:

>Heavy Fuel Oil – 74%
>• Kerosene – 17.4%
>• Bagasse – 5.6%
>• Diesel – 2.2%
>• Natural Gas – 0.2%
>• Solar – 0.6%
(It also notes that solar electricity increased to 2.4% in 2016, though we don't need that info as the data source returns solar data.)

The page also has the feedstock consumed in MMBTU (million British Thermal Units), which should allow for an even more exact calculation.

It also says "* Accounts for use of natural gas to produce electricity in the manufacturing sector.", which might imply [embedded generation](https://en.wikipedia.org/wiki/Captive_power_plant).
-----
Anyway, no more recent data than 2015

Reply:
I found some graphs in the 2021 "[MM_IADB_Final IRRP Report_Activity B – Energy.gov.bb](https://energy.gov.bb/download/mm_iadb_final-irrp-report_activity-b/) that suggest that, in 2021, Barbados only had oil-based and solar-based electricity.

![Figure 7.10 and 7.11 from the linked report showing two scenarios with identical installed capacity in 2021 as a starting point](https://github.com/electricitymaps/electricitymaps-contrib/assets/2470175/390a2b4d-df8d-4e2a-9ee8-9a19ffe6dcc6)

Reply:
These charts suggest that "fossil" is oil, right?

Reply:
Yes. My main remaining reservation is that it's still somewhat dated... I wish we had more recent data.

But I guess tracking it as oil is okay.